ls1intum/Artemis
# `Development`: Lazily load notifications when opening sidebar

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Improve the performance by only loading the amount of unread messages and lazily load notifications once the sidebar is opened.

### Description

This PR adds a new Rest-Endpoint that returns the amount of unread notifications. This is used to display the user if they have unread messages. The actual messages are only loaded once the notification sidebar is opened.
Additionally the lastNotificationRead is now only updated if it actually affects any notification. I.e. if there are no unread messages, the date is not updated via a rest call.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students

1. Log in to Artemis
2. Do something to receive a notification. E.g. wait until a new exercise is released.
3. Check that the bell icon shows unread messages when freshly loading the page.
4. Check that you can then see the notification in the side bar.
5. Check that the notification is marked as read and you no longer have unread notifications even when reloading.

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage





#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [notification-sidebar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4992-3/artifact/shared/Coverage-Report-Client-Tests/app/shared/notification/notification-sidebar/notification-sidebar.component.ts.html) | 87.66% | ✅ |
| [notification.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4992-3/artifact/shared/Coverage-Report-Client-Tests/app/shared/notification/notification.service.ts.html) | 84.37% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [NotificationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4992-3/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/NotificationResource.html) | 93% | ✅ |
